src/models/restricted_pairwise_union_reach_surface.py

```diff
@@ -16,127 +16,335 @@
+        """Compute B(lbd_{-i}) of equation (5) in the algorithm description doc.
+
+        B(lbd_{-i}) is the upper bound of lbd[i] so that lbd falls in the
+        feasibele region.  This is an intermediate method in self._step.
+
+        Args:
+          lbd:  The current best guess of lbd.
+          i:  The index of the coordinate to update
+          tol:  An artifical threshold to avoid divide-by-zero error.
 
         Returns:
-          the list of constraint functions
+          A bound B satisfying the following property.  Suppose we change lbd[i]
+          to a number 'l_new' while keeping the other coordinates unchanged.
+          Then lbd is feasible (i.e., cls._check_lbd_feasiblity == True) if and
+          only if 0 <= l_new <= B.
         """
+        total = sum(lbd)
+        res = 1 / max(total - lbd[i], 1e-6)
+        for j in range(len(lbd)):
+            if j != i:
+                res = min(res, 1 / max(lbd[j], tol) - total + lbd[i] + lbd[j])
+        return res
 
-        cons = []
-        for i in range(self._p):
-            # All lambdas are non negative : lbd[i] >= 0
-            cons.append({"type": "ineq", "fun": lambda x: x[i]})
-            # Lambda j sum times Lambda i is less than 1 : 1 - lbd[i] * sum(lbd) >= 0
-            cons.append({"type": "ineq", "fun": lambda x: 1 - x[i] * sum(x)})
-        return cons
+    def _round(self, lbd: List[float]):
+        """Each round to update the whole vector lbd.
 
-    def _evaluate_point(self, lbd: List[float], reach_vector: List[float]):
-        """Evaluate reach at a point.
+        A round consists of self._p steps.
 
         Args:
-          lbd: a length p vector indicating lambda_i for each pub.
-          reach_vector: a length p vector indicating the single-pub reach of each
-            pub at a single data point.
+          lbd:  The current best guess of lbd.
 
         Returns:
-          the value of union reach at the given point.
+          The updated best guess of lbd after a round.
         """
-        reach_sum = sum(reach_vector)
-        overlap = sum(
-            [
-                (lbd[i] * lbd[j] * reach_vector[i] * reach_vector[j])
-                / (
-                    max(
-                        self._reach_curves[i].max_reach, self._reach_curves[j].max_reach
-                    )
-                    * 2
-                )
-                for i in range(self._p - 1)
-                for j in range(i, self._p)
-            ]
-        )
-        return reach_sum - overlap
+        for i in np.random.permutation(self._p):
+            # We shuffle the order of coordinates to have symmetry across
+            # different EDPs.
+            lbd[i] = self._step(lbd, i)
+        return lbd
 
     def _loss(self, lbd: List[float]):
-        """Get value of loss function.
+        """Compute the L2 loss of any lbd."""
+        loss = 0
+        for y, D in zip(self._ys, self._Ds):
```

Comment:
This seems like it could also be expressed as a matrix operation.

Reply:
Will try